File: runtime/perl/units/table/call.pl
# requires: rt/trap
sub call {
    my ($self, $i, $type_key, @args) = @_;
    Rt::trap('undefined element') if $i >= scalar @{$self->{slots}};
    my $slot = $self->{slots}[$i];
    Rt::trap('uninitialized element') unless defined $slot;
    Rt::trap('indirect call type mismatch') unless $slot->[0] eq $type_key;
    return $slot->[1]->(@args);
}

File: runtime/go/units/table/call.go
// requires: rt/trap
// The caller type-asserts the returned func value to the exact signature of the
// call site.
func (t *Table) call(i uint32, typeKey string) any {
    if i >= uint32(len(t.slots)) {
        Rt.trap("undefined element")
    }
    slot := t.slots[i]
    if slot == nil {
        Rt.trap("uninitialized element")
    }
    if slot.ty != typeKey {
        Rt.trap("indirect call type mismatch")
    }
    return slot.fn
}
